Editorial review: Document new web speech api features (#41145)
mdn/content: files/en-us/web/api/speechrecognition/available_static/index.md

---
title: "SpeechRecognition: available() static method"
short-title: available()
slug: Web/API/SpeechRecognition/available_static
page-type: web-api-static-method
browser-compat: api.SpeechRecognition.available
---

{{APIRef("Web Speech API")}}

> The **`available()`** static method of the [Web Speech API](/en-US/docs/Web/API/Web_Speech_API) checks whether specified languages are available for speech recognition either locally on the user's computer, or via a remote service.

Author: OK, I think I get the issue here — the initial description was a little bit ambiguous. I've cut it down to just say "...checks whether specified languages are available for speech recognition", and then made sure that I explain clearly what the `processLocally` option does in the Parameters description below. I added a note to clarify that you can't use `available()` to determine whether a remote service is guaranteed to support the specified languages.

Let me know what you think of the updates.

> The **`available()`** static method of the [Web Speech API](/en-US/docs/Web/API/Web_Speech_API) checks whether specified languages are available for speech recognition either locally on the user's computer, or via a remote service.

Reviewer: Looks good to me!

Access to the `available()` method is controlled by the {{httpheader("Permissions-Policy/on-device-speech-recognition", "on-device-speech-recognition")}} {{httpheader("Permissions-Policy")}} directive. Specifically, where a defined policy blocks usage, any attempts to call the method will fail.

## Syntax

```js-nolint
available(options)
```

### Parameters

- `options`
  - : An object specifying options for the availability check. Possible properties include:
    - `langs`
      - : An array of one or more strings containing {{glossary("BCP 47 language tag", "BCP 47 language tags")}}, each representing a language that you want to check availability of for speech recognition. Passing an empty `langs` array will not throw an error, but the return value will always resolve to `unavailable`.
    - `processLocally` {{optional_inline}}
      - : A boolean that specifies whether you are checking availability of the specified languages for [on-device speech recognition](/en-US/docs/Web/API/Web_Speech_API/Using_the_Web_Speech_API#on-device_speech_recognition) (`true`) or availability of the specified languages for on-device or remote speech recognition (`false`). Defaults to `false`.

> ### Return value

Reviewer: It might be a good idea to document what happens when multiple languages are specified with different availability: https://webaudio.github.io/web-speech-api/#availability-algorithm

Author: Can you clarify what you think needs explaining?

In the description of the different possible return values, I've explained that `available` is returned if support for all the languages is available, and `unavailable` is returned if at least one language is not supported. 

I've done a bit of restructuring to make this a bit clearer. Am I misunderstanding something?

Reviewer: The available() method takes in a collection of languages, so if a user calls it with en-US and fr-FR for example, en-US might be "available" but fr-FR might be "downloading", so the method would resolve with "downloading" since it only returns a single status for all of the languages specified. The section on the availability algorithm in the spec describes this behavior.

A {{domxref("Promise")}} that resolves with an emumerated value indicating the availability of the specified languages for speech recognition. Possible values are:

- `available`

>   - : Indicates that the specified languages are available. If `processLocally` is set to `true`, `available` means that the required language packs have been downloaded and installed on the user's computer. If `processLocally` is set to `false`, `available` means that speech recognition is availale for those languages either on-device or remotely.

Reviewer: availale->available

Author: Good spot; fixed!

The **`available()`** static method of the [Web Speech API](/en-US/docs/Web/API/Web_Speech_API) checks whether the specified languages are available for speech recognition.

To install a language pack for speech recognition locally, you need to use the {{domxref("SpeechRecognition.install_static", "SpeechRecognition.install()")}} method.

Access to the `available()` method is controlled by the {{httpheader("Permissions-Policy/on-device-speech-recognition", "on-device-speech-recognition")}} {{httpheader("Permissions-Policy")}} directive. Specifically, where a defined policy blocks usage, any attempts to call the method will fail.

## Syntax

```js-nolint
available(options)
```

### Parameters

- `options`
  - : An object specifying options for the availability check. Possible properties include:
    - `langs`
      - : An array of one or more strings containing {{glossary("BCP 47 language tag", "BCP 47 language tags")}}, each representing a language that you want to check availability of for speech recognition. Passing an empty `langs` array will not throw an error, but the return value will always resolve to `unavailable`.
    - `processLocally` {{optional_inline}}

>       - : A boolean that specifies whether you are checking availability of the specified languages for [on-device speech recognition](/en-US/docs/Web/API/Web_Speech_API/Using_the_Web_Speech_API#on-device_speech_recognition) (`true`) or availability of the specified languages for on-device _or_ remote speech recognition (`false`). Defaults to `false`.

Reviewer 2: ```suggestion
      - : A boolean that specifies whether to check availability of the languages only for [on-device speech recognition](/en-US/docs/Web/API/Web_Speech_API/Using_the_Web_Speech_API#on-device_speech_recognition) (`true`) or for on-device _or_ remote speech recognition (`false`). The default value is `false`.
```

        > [!NOTE]

>         > It is not possible to use `available()` to determine whether a remote service is guaranteed to support the specified languages. A value of `false` means that either an on-device _or_ remote speech recognition service supports the specified languages.

Reviewer 2: ```suggestion
        > It is not possible to use `available()` to guarantee that a remote service supports the specified languages. A value of `false` means that either an on-device _or_ a remote speech recognition service supports them.
```

### Return value

A {{domxref("Promise")}} that resolves with an enumerated value indicating the availability of the specified languages for speech recognition.

> Possible values are as follows.

Reviewer 2: ```suggestion
Possible values include:
```

    - If `processLocally` is set to `false`, `available` means that speech recognition is available for those languages either on-device or remotely.
- `downloading`
  - : Indicates that support for the specified languages is available on-device, and the relevant language pack for at least one language is in the process of being downloaded. Only relevant when `processLocally` is `true`.
- `downloadable`
  - : Indicates that support for the specified languages is available on-device, but the relevant language pack for at least one language has not yet been downloaded. Only relevant when `processLocally` is `true`.
- `unavailable`
  - : Indicates that support for at least one of the specified languages is not available.
    - If `processLocally` is set to `true`, `unavailable` means that on-device speech recognition is not available for at least one of the specified languages.
    - If `processLocally` is set to `false`, `unavailable` means that speech recognition is not available for at least one of the specified languages either on-device or remotely.

#### Final return value for multiple languages with different statuses

> Only a single status value is returned, even if multiple languages are specified in the `langs` array. If different specified languages have different availability statuses, the final returned status will be the "furthest away" status to `available` of any of the languages, in the order shown in the following lists.

Reviewer 2: ```suggestion
Only one status value is returned, even if multiple languages are specified in the `langs` array. If different specified languages have different availability statuses, the final return value is the "furthest away" status from `available` for any of the languages, in the order shown in the following lists:
```

If `processLocally` is `false`:

- If all languages are `available`, then return `available`.
- Otherwise, return `unavailable`.

If `processLocally` is `true`:

- If all languages are `available`, return `available`.
- If at least one language is `downloading`, return `downloading`.
- If at least one language is `downloadable`, return `downloadable`.
- If at least one language is `unavailable`, return `unavailable`.

### Exceptions

- `InvalidStateError` {{domxref("DOMException")}}
  - : The current document is not fully active.
- `SyntaxError` {{domxref("DOMException")}}
  - : One or more of the strings specified in `langs` is not a valid BCP 47 language tag.

## Examples

> ### Checking on-device availability and installing language packs

Reviewer 2: i believe the text in this section is similar to the one in the "Using" page and so same edits can be applied here
